app/services/broadcasts/welcome_notification/generator.rb

```diff
@@ -11,27 +10,51 @@ def self.call(*args)
       end
 
       def call
-        return if commented_on_welcome_thread? || received_notification?
+        send_welcome_notification
+        send_authentication_notification
+      end
+
+      def send_welcome_notification
+        return if received_notification?(welcome_broadcast) || commented_on_welcome_thread? || user.created_at > 3.hours.ago
 
         Notification.send_welcome_notification(user.id, welcome_broadcast.id)
       end
 
-      def received_notification?
-        Notification.exists?(notifiable: welcome_broadcast, user: user)
+      def send_authentication_notification
+        return if authenticated_with_both_services? || received_notification?(authentication_broadcast) || user.created_at > 1.day.ago
+
+        Notification.send_welcome_notification(user.id, authentication_broadcast.id)
+      end
+
+      private
+
+      attr_reader :user
+
+      def received_notification?(broadcast)
+        Notification.exists?(notifiable: broadcast, user: user)
       end
 
       def commented_on_welcome_thread?
         welcome_thread = Article.admin_published_with("welcome").first
         Comment.where(commentable: welcome_thread, user: user).any?
       end
 
-      private
+      def authenticated_with_both_services?
+        identities.count == 2
+      end
 
       def welcome_broadcast
         @welcome_broadcast ||= Broadcast.find_by(title: "Welcome Notification: welcome_thread")
       end
 
-      attr_reader :user
+      def identities
+        @identities ||= user.identities.where(provider: %w[github twitter])
+      end
+
+      def authentication_broadcast
+        missing_identity = identities.exists?(provider: "github") ? "twitter_connect" : "github_connect"
```

Comment:
@vaidehijoshi Re:
> I think this logic wouldn't be quite right if someone had authenticated with both services, right? 🤔

So I created instance var `identities` that can be used in multiple areas. I think it would be a bit much to also double check here if user has two identities since that will be done in `authenticated_with_both_services?` with the same variable. I'm not sure if this is the best way but I would rely on the spec to catch the edge-case.  Let me know what you think.

Reply:
Code outdated.